### PR TITLE
Mqtt Support

### PR DESCRIPTION
**Pull request by ihong4**
pub and sub mqtt client

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -1,91 +1,102 @@
...
+use std::{
+    env,
+    process,
+    thread,
+    time::Duration
+};
+
+extern crate paho_mqtt as mqtt;
+
+const DFLT_BROKER:&str = "tcp://localhost:1883";
```
Comment (Peyton-McKee) [inline src/main.rs:10]: what is DFLT

File: src/main.rs [hunk 2]
```diff
@@ -1,91 +1,102 @@
...
-use std::env;
-use std::io::Write;
-use std::os::unix::net::UnixStream;
-use std::process::Command;
-use std::sync::mpsc::channel;
-use std::thread;
-mod data;
-mod decode_data;
-mod master_mapping;
-mod message;
```
Comment (Peyton-McKee) [inline src/main.rs:15]: why are you deleting all this

File: src/main.rs [hunk 3]
```diff
@@ -1,91 +1,102 @@
...
+extern crate paho_mqtt as mqtt;
+
+const DFLT_BROKER:&str = "tcp://localhost:1883";
+const DFLT_CLIENT:&str = "rust_subscribe";
+const DFLT_TOPICS:&[&str] = &["/mpu", "/tpu"];
+// The qos list that match topics above.
+const DFLT_QOS:&[i32] = &[0, 1];
+
+// Reconnect to the broker when connection is lost.
+fn try_reconnect(cli: &mqtt::Client) -> bool
```
Comment (Peyton-McKee) [inline src/main.rs:17]: this shoudlnt be in main file

File: src/main.rs [hunk 4]
```diff
@@ -1,91 +1,102 @@
...
+            println!("Successfully reconnected");
+            return true;
+        }
+    }
+    println!("Unable to reconnect after several attempts.");
+    false
+}
+
+// Subscribes to multiple topics.
+fn subscribe_topics(cli: &mqtt::Client) {
```
Comment (Peyton-McKee) [inline src/main.rs:32]: same with this

File: src/main.rs [hunk 5]
```diff
@@ -1,91 +1,102 @@
...
+        println!("Error subscribes topics: {:?}", e);
+        process::exit(1);
+    }
+}
 
 fn main() {
-    let args: Vec<String> = env::args().collect();
-    let default = "tmp/ipc.sock".to_owned();
-    let ipc_path = args.get(0).unwrap_or(&default);
+    let host = env::args().nth(1).unwrap_or_else(||
```
Comment (Peyton-McKee) [inline src/main.rs:40]: this is not correct. Calypso is a shared application, NERO also uses this file, so you have to figure out what application is using it from the command line args. You need to bring back the other stuff

File: src/main.rs [hunk 6]
```diff
@@ -1,91 +1,102 @@
...
-        .expect("Fail while waiting for bit rate");
-    let mut up_command = Command::new("sudo")
-        .arg("ifconfig")
-        .arg("can0")
-        .arg("up")
-        .spawn()
-        .expect("up command did nto work");
-    up_command
-        .wait()
-        .expect("Fail while waiting for up command");
```
Comment (Peyton-McKee) [inline src/main.rs:53]: why did you delete all of this. i think this will still be used in order to communicate with the can hat.

File: src/main.rs [hunk 7]
```diff
@@ -1,91 +1,102 @@
...
+    // Define the set of options for the create.
+    // Use an ID for a persistent session.
+    let create_opts = mqtt::CreateOptionsBuilder::new()
```
Comment (Peyton-McKee) [inline src/main.rs:46]: this should be in a separate function that gets called if the command line args specify for mqtt or NERO

File: src/main.rs [hunk 8]
```diff
@@ -1,91 +1,102 @@
...
+        .server_uri(host)
+        .client_id(DFLT_CLIENT.to_string())
+        .finalize();
 
-    let mut stream = UnixStream::connect(ipc_path).unwrap();
```
Comment (Peyton-McKee) [inline src/main.rs:55]: why are you deleting this, you still need the can socket in order to actually get the information off the car

File: src/main.rs [hunk 9]
```diff
@@ -1,91 +1,102 @@
...
-    //open can socket channel at name can0
-    const CAN_CHANNEL: &str = "can0";
-    let socket = CANSocket::open(&CAN_CHANNEL);
-    let socket = match socket {
-        Ok(socket) => socket,
-        Err(err) => {
-            println!("Failed to open CAN socket: {}", err);
-            return;
+    // Create a client.
+    let mut cli = mqtt::Client::new(create_opts).unwrap_or_else(|err| {
```
Comment (Peyton-McKee) [inline src/main.rs:52]: all of this should be in a seperate file and should just be called from the main function

File: src/main.rs [hunk 10]
```diff
@@ -1,91 +1,102 @@
...
+    // Subscribe topics.
+    subscribe_topics(&cli);
+
+    println!("Processing requests...");
+    for msg in rx.iter() {
+        if let Some(msg) = msg {
+            println!("{}", msg);
         }
-    };
-    thread::spawn(move || loop {
```
Comment (Peyton-McKee) [inline src/main.rs:67]: this shoudlnt be deleted. It should be moved to a seperate file called like NERO Socket

**Review by Peyton-McKee — CHANGES_REQUESTED** [review]
while this may work for connecting to mqtt. you still need the can decoding process which you have deleted. I recommend creating two new files, one called NERO socket and the other one called MQTT Socket, put all the mqtt code int he mqtt socket file and all the nero code in the nero socket file. Then parse the command line arguments and call the top function for each depending on what the command line args are.

File: src/main.rs [hunk 11]
```diff
@@ -64,28 +56,57 @@ fn main() {
             return;
         }
     };
-    thread::spawn(move || loop {
```
Comment (evanlombardo) [inline src/main.rs:67]: any reason you took out the threading?

Comment (Peyton-McKee) [reply src/main.rs:67]: it wasnt working, debugging rn might add it back later

File: oxy/structs/Format.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+class Timer(Format):
+      repr: str = "timer"
+
+@dataclass
+class Flux(Format):
+      repr: str = "flux"
+
+@dataclass
+class CellVoltage(Format):
+      repr: str = "cell_voltage"
```
Comment (RChandler234) [inline oxy/structs/Format.py:56]: you've got all these newline indicators, not sure if you care ab that tho

**Review by RChandler234 — CHANGES_REQUESTED** [review]

File: Cargo.lock [hunk 13]
```diff
@@ -2,6 +2,12 @@
 # It is not intended for manual editing.
```
Comment (nwdepatie) [inline Cargo.lock:2]: I've found conflicting thoughts on this online: should we be ignoring `cargo.lock`?

Comment (nwdepatie) [reply Cargo.lock:2]: Very large diff here and not sure if necessary

File: oxy/yaml/bms.yaml [hunk 14]
```diff
@@ -0,0 +1,243 @@
+!Messages
```
Comment (nwdepatie) [inline oxy/yaml/bms.yaml:1]: Nit: can we rename the `yaml` directory to something like `msgs` or something else that describes what is contained in the yaml files?

File: oxy/yaml/bms.yaml [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+    - !CANField
+        name: "BMS/Cells/Volts/Low/Cell"
+        index: 5
+        size: 1
+        unit: ""
+        decodings:
+        - !Half
+            bits: 0
+    - !CANField
+        name: "BMS/Cells/Volts/Ave/Value"
```
Comment (nwdepatie) [inline oxy/yaml/bms.yaml:138]: Nit: I prefer `Avg` vs `Ave` for average. If its not average then idk what it is lol

File: oxy/yaml/mpu.yaml [hunk 16]
```diff
@@ -0,0 +1,51 @@
+!Messages
+msgs:
+- !CANMsg
+    id: "0x500"
+    desc: "NERduino Acceleromter"
```
Comment (nwdepatie) [inline oxy/yaml/mpu.yaml:5]: Don't call this NERduino anymore, we aren't using that board

File: oxy/yaml/mpu.yaml [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+        unit: "g"
+        size: 2
+        decodings: 
+        - !BigEndian
+            bits: 8
+        format: "acceleration"
+
+- !CANMsg
+    id: "0x501"
+    desc: "MPU Status"
```
Comment (nwdepatie) [inline oxy/yaml/mpu.yaml:34]: Nit: I know that this is the MPU, but I think we might wanna rename to Ceberus Status or something because I think the RTOS application might outlive the actual PCB it is on

Comment (nwdepatie) [reply oxy/yaml/mpu.yaml:34]: Actually jk this is prob a slightly bigger discussion

File: oxy/yaml/wheel.yaml [hunk 18]
```diff
@@ -0,0 +1,14 @@
+!Messages
+msgs:
+- !CANMsg
+    id: "0x680"
+    desc: "Wheel State"
+    fields: 
+    - !CANField
+        name: "WHEEL/Buttons/1"
```
Comment (nwdepatie) [inline oxy/yaml/wheel.yaml:8]: Uh doesn't the wheel message broadcast more than just 2 buttons? Isn't it bitwise encoded with like 10 this time around? Or are we waiting for wheel assembly/integration to fix this up?

File: src/data.rs [hunk 19]
```diff
@@ -1,142 +1,152 @@
...
 impl FormatData {
-    pub fn temperature(value: i64) -> f32 {
-        value as f32 / 10.0
+    pub fn temperature(value: f32) -> f32 {
+        value / 10.0
     }
 
-    pub fn low_voltage(value: i64) -> f32 {
-        value as f32 / 100.0
+    pub fn low_voltage(value: f32) -> f32 {
```
Comment (nwdepatie) [inline src/data.rs:105]: Can you elaborate on what "low voltage" and "high voltage" entail? Is it like actually a voltage value above around 50V? Is it mV vs V? Is it just a higher resolution value?

File: src/data.rs [hunk 20]
```diff
@@ -1,142 +1,152 @@
...
 
-    pub fn power(value: i32) -> f32 {
-        value as f32 / 10.0
+    pub fn flux(value: f32) -> f32 {
+        value / 1000.0
     }
 
-    pub fn timer(value: i32) -> f32 {
-        value as f32 * 0.003
+    pub fn cell_voltage(value: f32) -> f32 {
```
Comment (nwdepatie) [inline src/data.rs:145]: Why do we need a different format for cell voltage? Just because we broadcast with a higher resolution and need to normalize differently?

File: src/data.rs [hunk 21]
```diff
@@ -1,142 +1,152 @@
...
+        value / 10000.0
     }
 
-    pub fn flux(value: i64) -> f32 {
-        value as f32 / 1000.0
+    pub fn acceleration(value: f32) -> f32 {
+        value * 0.0029
```
Comment (nwdepatie) [inline src/data.rs:150]: Where is this 0.0029 number coming from?

File: src/data.rs [hunk 22]
```diff
@@ -1,142 +1,152 @@
...
 
-pub struct FormatData {
-    // Utility functions to scale data values of a specific type.
-}
+/**
+ * Class to contain the data formatting functions
+ */
+pub struct FormatData {}
 
 impl FormatData {
```
Comment (nwdepatie) [inline src/data.rs:100]: I think that these FormatData functions are a bit ambiguous and use a lot of magic numbers. We should probably add either a comment explaining where the normalization factor is coming from and what type of data should be used with it

File: src/main.rs [hunk 23]
```diff
@@ -1,31 +1,19 @@
...
 
-fn main() {
-    let args: Vec<String> = env::args().collect();
-    let default = "tmp/ipc.sock".to_owned();
-    let ipc_path = args.get(0).unwrap_or(&default);
+use calypso::{client::Client, message::Message, mqtt::MqttClient};
+use socketcan::CANSocket;
 
+fn configure_can() {
     let mut down_command = Command::new("sudo")
```
Comment (nwdepatie) [inline src/main.rs:10]: Nit: is there any way we can sidestep using sudo here? Feels weird for an application to be using a sudo command when being deployed when we could do something with user groups or something. Could also be necessary so just food for thought

File: src/main.rs [hunk 24]
```diff
@@ -48,44 +36,80 @@ fn main() {
...
-    let mut stream = UnixStream::connect(ipc_path).unwrap();
-    let (tx, rx) = channel();
+/**
+ * Reads the can socket and publishes the data to the given client.
+ */
+fn read_can(mut publisher: Box<dyn Client + Send>) {
     //open can socket channel at name can0
     const CAN_CHANNEL: &str = "can0";
-    let socket = CANSocket::open(&CAN_CHANNEL);
+    let socket = CANSocket::open(CAN_CHANNEL);
```
Comment (nwdepatie) [inline src/main.rs:50]: Why the removal of the &? Have we tested this?

**Review by nwdepatie — COMMENTED** [review]
I left some comments to be addressed, mostly nitpicks but some general questions

File: Cargo.lock (hunk 13 above)
Comment (Peyton-McKee) [reply Cargo.lock:2]: It doesnt particularly matter its just our dependencies. We keep our yarn.lock and package-lock.json in our other projects so we might as well keep it here. Ideally if we dont chnage our dependencies this wont change either.

File: oxy/yaml/mpu.yaml (hunk 16 above)
Comment (Peyton-McKee) [reply oxy/yaml/mpu.yaml:5]: What should i call it

File: oxy/yaml/wheel.yaml (hunk 18 above)
Comment (Peyton-McKee) [reply oxy/yaml/wheel.yaml:8]: So what we do currently is have each button be one bit and then we decode it on the nero side. Now that we have ten buttons it heeds to he two messages for 10 bits

File: src/data.rs (hunk 19 above)
Comment (Peyton-McKee) [reply src/data.rs:105]: Honestly thats legacy from peter idrk

File: src/data.rs (hunk 20 above)
Comment (Peyton-McKee) [reply src/data.rs:145]: Yeah its dividing by 1000 whe i think the other voltage formats are only by ten

File: src/data.rs (hunk 21 above)
Comment (Peyton-McKee) [reply src/data.rs:150]: The accelerometer data dheet i believe

File: src/data.rs (hunk 22 above)
Comment (Peyton-McKee) [reply src/data.rs:100]: I agree the issue is that most of these are legacy from peter so i dont have the best idea on what the majority are

File: src/main.rs (hunk 24 above)
Comment (Peyton-McKee) [reply src/main.rs:50]: None of this has been tested but its changed anyways in jacks pr

File: src/main.rs (hunk 23 above)
Comment (Peyton-McKee) [reply src/main.rs:10]: I dont know thatd more of a jack linux question. Hubert made this code i believe

File: oxy/yaml/mpu.yaml (hunk 16 above)
Comment (nwdepatie) [reply oxy/yaml/mpu.yaml:5]: Prob MPU Accelerometer

File: oxy/structs/Format.py (hunk 12 above)
Comment (Peyton-McKee) [reply oxy/structs/Format.py:56]: eh not really

**Comment by Peyton-McKee** [comment]
Will be Continued in #30
